devscripts/make_changelog.py

```diff
@@ -0,0 +1,452 @@
+                self._commits[commit.hash] = commit
+
+        self._commits = {key: value for key, value in reversed(self._commits.items())}
+
+    def groups(self):
+        groups = defaultdict(list)
+        for commit in self:
+            match = self.MESSAGE_RE.fullmatch(commit.short)
+            if not match:
+                logger.error(f'Error parsing short commit message: {commit.short!r}')
+                continue
+
+            prefix, details, sub_details, sub_details_alt, message, issues = match.groups()
+            group = None
+            if prefix:
+                if prefix == 'priority':
+                    prefix, _, details = (details or '').partition('/')
+                    logger.debug(f'Priority: {message!r}')
+                    group = CommitGroup.PRIORITY
+
+                if not details and prefix:
+                    if prefix not in ('extractor', 'postprocessor', 'downloader'):
+                        logger.debug(f'Replaced details with {prefix!r}')
+                        details = prefix or None
+
+                if details == 'common':
+                    details = None
+
+                if details:
+                    details = details.strip()
+
+            else:
+                group = CommitGroup.CORE
+
+            sub_details = f'{sub_details or ""},{sub_details_alt or ""}'.lower().replace(':', ',')
+            sub_details = tuple(filter(None, map(str.strip, sub_details.split(','))))
+
+            issues = [issue.strip()[1:] for issue in issues.split(',')] if issues else []
+
+            if not group:
+                group = CommitGroup.get(prefix.lower())
+                if not group:
+                    if self.EXTRACTOR_INDICATOR_RE.search(commit.short):
+                        group = CommitGroup.EXTRACTOR
+                    else:
+                        group = CommitGroup.POSTPROCESSOR
+                    logger.warning(f'Failed to map {commit.short!r}, selected {group.name}')
+
+            commit_info = CommitInfo(details, sub_details, message.strip(), issues, commit)
+            logger.debug(f'Resolved {commit.short!r} to {commit_info!r}')
+            groups[group].append(commit_info)
+
+        return groups
+
+
+def get_new_contributors(contributors_path, commits):
+    contributors = set()
+    if contributors_path.exists():
+        with contributors_path.open() as file:
+            for line in filter(None, map(str.strip, file)):
+                author, _, _ = line.partition(' (')
+                authors = author.split('/')
+                contributors.update(map(str.lower, authors))
+
+    new_contributors = {}
+    for commit in commits:
+        for author in commit.authors:
+            author_compare = author.lower()
+            if author_compare in contributors:
+                continue
+            contributors.add(author_compare)
+            new_contributors[author] = None
+
+    return list(reversed(new_contributors))
```

Comment:
Why use a dict?

```suggestion
    return sorted({
        author for commit in commits for author in commit.authors
        if author.casefold() not in contributors
    })
```

Reply:
I want to keep the order in which the contributors appeared and not sort them by name. Its fine to not do this though if the order does not matter.

Reply:
The current list is not sorted in order of commit

Reply:
This inlining doesnt quite work if casing changed for new contributors (`Grub4K`, `Grub4k`), though it could be argued that that should be corrected using changelog override. If it should I will use this variant, however, as this type of mistake is easily made, it might be better to implement the autocorrection.